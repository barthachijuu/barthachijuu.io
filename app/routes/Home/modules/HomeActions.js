/*
 *
 * Home Actions
 *
 * @version 1.0.0
 * @author Bartolo Amico
 * @since 13-Jul-2020
 */

import { createRequestTypes, createAction, MakeRequest } from 'Store/actions/index';

export const actionTypes = {
  GET_INFO: 'home/get_info',
  OPEN_MODAL: 'home/open_modal',
  CLOSE_MODAL: 'home/close_modal',
  // @generator action:route:type
  ...createRequestTypes('home', 'GET_INFO'),
// @generator action:route:type:request
};

// ------------------------------------
// Actions
// ------------------------------------
// @generator action:route:request
/** @action request
* * Insert a description
* @name getInfo
* @return {object} - The action object
*/
export const getInfo = new MakeRequest('home/get_info');

// @generator action:route
/** @action
* * Insert a description
* @name CloseModal
* @return {object} - The payload object
*/
export const doCloseModal = payload => createAction(actionTypes.CLOSE_MODAL, payload);
/** @action
* * Insert a description
* @name OpenModal
* @return {object} - The payload object
*/
export const doOpenModal = payload => createAction(actionTypes.OPEN_MODAL, payload);
/** @action
* * Insert a description
* @name GetInfo
* @return {object} - The payload object
*/
export const doGetInfo = payload => createAction(actionTypes.GET_INFO, payload);
