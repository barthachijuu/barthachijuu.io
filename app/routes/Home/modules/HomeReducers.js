/*
 * Home Reducer
 * Description
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 * @version 1.0.0
 * @author Bartolo Amico
 * @since 13-Jul-2020
 */

import produce from 'immer';
import { actionTypes } from './HomeActions';

const ACTION_HANDLERS = {
  [actionTypes.GET_INFO_SUCCESS]: produce((draft, action) => {
    draft.info = action.payload.response.data;
  }),
  [actionTypes.GET_INFO_FAILURE]: produce((draft) => {
    draft.info = undefined;
  }),
  [actionTypes.OPEN_MODAL]: produce((draft, action) => {
    draft.info.projects[action.payload.idx].isOpen = action.payload.status;
  }),
  [actionTypes.CLOSE_MODAL]: produce((draft, action) => {
    draft.info.projects[action.payload.idx].isOpen = action.payload.status;
  }),
  // @generator reducer:type:action

};

// The initial state of the home
const initialState = {
  info: {},
};

const home = (state = initialState, action) => {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
};

export default home;
