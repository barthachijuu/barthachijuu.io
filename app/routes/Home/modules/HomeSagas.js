/*
 *
 * Home Sagas
 *
 * @version 1.0.0
 * @author Bartolo Amico
 * @since 13-Jul-2020
 */

import { put, take, call, all } from 'redux-saga/effects';
import { doStatusSuccess, doStatusFailure } from 'Store/actions/enhancers';
import * as api from 'Api/home/home';
import * as actions from './HomeActions';

// remember to delete the example code, before you
// start to edit this file


export function* retrieveInfo() {
  yield put(actions.getInfo.request());
  const { response, error } = yield call(api.getInfo);
  if (response) {
    yield put(actions.getInfo.success(response));
    yield put(doStatusSuccess());
  } else {
    yield put(actions.getInfo.failure(error));
    yield put(doStatusFailure());
  }
}
// @generator saga

// @generator saga:watch

export function* watchRetrieveInfo() {
  while (yield take(actions.actionTypes.GET_INFO)) {
    yield call(retrieveInfo);
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* rootSaga() {
  yield all([
    watchRetrieveInfo(),
    // @generator saga:fork

  ]);
}
