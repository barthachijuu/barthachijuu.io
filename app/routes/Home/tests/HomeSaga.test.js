/**
 *
 * Jest Unit Test for
 * Home Redux Modules
 *
 */

import { take, put } from 'redux-saga/effects';
import { retrieveInfo, watchRetrieveInfo } from '../modules/HomeSagas';
import * as actions from '../modules/HomeActions';


describe('get info saga', () => {
  let retrieveInfoGenerator;

  beforeEach(() => {
    retrieveInfoGenerator = retrieveInfo();

    const selectDescriptor = retrieveInfoGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = retrieveInfoGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch action "GET_INFO" ', () => {
    const generator = watchRetrieveInfo();
    expect(generator.next().value).toEqual(take('home/get_info', retrieveInfo));
    expect(generator.next().done).toBeTruthy();
  });

  it('should dispatch action "GET_INFO_REQUEST"', () => {
    const gen = retrieveInfo();
    const request = { payload: {} };
    const putDescriptor = gen.next(request).value;
    expect(putDescriptor).toEqual(put({ type: actions.actionTypes.GET_INFO_REQUEST, payload: {} }));
  });

  it('should dispatch action "GET_INFO_SUCCESS"', () => {
    const gen = retrieveInfo({});
    const payload = { response: { } };
    gen.next();
    gen.next();
    expect(gen.next(payload).value).toEqual(put({ type: actions.actionTypes.GET_INFO_SUCCESS, payload }));
  });

  it('should dispatch action "GET_INFO_FAILURE"', () => {
    const gen = retrieveInfo({});
    const payload = { error: { } };
    gen.next();
    gen.next();
    expect(gen.next(payload).value).toEqual(put({ type: actions.actionTypes.GET_INFO_FAILURE, payload }));
    expect(gen.next().done).toBeFalsy();
  });
});
