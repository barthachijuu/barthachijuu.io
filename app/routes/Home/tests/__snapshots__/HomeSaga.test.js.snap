// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get info saga should dispatch action "GET_INFO"  1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "request": undefined,
      },
      "type": "home/get_info_request",
    },
    "channel": null,
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO"  2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_FAILURE" 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "request": undefined,
      },
      "type": "home/get_info_request",
    },
    "channel": null,
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_FAILURE" 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_REQUEST" 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "request": undefined,
      },
      "type": "home/get_info_request",
    },
    "channel": null,
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_REQUEST" 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_SUCCESS" 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "request": undefined,
      },
      "type": "home/get_info_request",
    },
    "channel": null,
  },
}
`;

exports[`get info saga should dispatch action "GET_INFO_SUCCESS" 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
